.github/workflows/on-push.yaml

```diff
@@ -0,0 +1,254 @@
+
+  # Runs the unit tests for the project (Rust + pnpm). It depends on the
+  # `build-tests` job to build the Nextest test archives and upload them as
+  # artifacts. Note that since we are using nextest archives, we do not need
+  # Rust to be installed in this job.
+  unit-tests:
+    name: Run Unit Tests
+    runs-on: ubuntu-24.04
+    needs: build-tests
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          submodules: "recursive"
+      - uses: pnpm/action-setup@v4
+      - uses: actions/setup-node@v3
+        with:
+          node-version: ${{ env.NODE_VERSION }}
+          cache: "pnpm"
+      - name: Install Nextest
+        run: curl -LsSf https://get.nexte.st/latest/linux | tar zxf - -C /usr/local/bin
+      - name: Install Package Dependencies
+        run: make install
+      - name: Download Nextest Archives
+        uses: actions/download-artifact@v4
+        with:
+          name: nextest-archives
+          path: .
+      - name: Run Unit Tests (Node)
+        run: pnpm --recursive test
+      - name: Run Unit Tests (Rust)
+        run: cargo-nextest nextest --color always run --no-fail-fast --retries 2 --archive-file nextest-archive.tar.zst
+
+  # Runs the Rust integration tests for the project. It depends on the
+  # `build-tests` job to build the nextest test archives and upload them as
+  # artifacts, however we do not define an explicit dependency with `needs`.
+  # This is because in this job, we also need to get the integration environment
+  # running, which generally takes around a minute. So, we start this job
+  # immediately and once the environment is running we use the `wait-other-jobs`
+  # action to wait until the `build-tests` job is complete and the artifacts are
+  # available. This lets us start the environment while the tests are building,
+  # and then run the tests as soon as the artifacts are available.
+  integration-tests:
+    name: Run Integration Tests
+    runs-on: ubuntu-24.04
+    strategy:
+      matrix:
+        partition: [1, 2]
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          submodules: "recursive"
+      - uses: actions-rust-lang/setup-rust-toolchain@v1
+        with:
+          cache-key: "rust-tests"
+      - uses: pnpm/action-setup@v4
+      - uses: actions/setup-node@v3
+        with:
+          node-version: ${{ env.NODE_VERSION }}
+          cache: "pnpm"
+      - uses: arduino/setup-protoc@v3
+        with:
+          version: "25.3"
+      - uses: actions/setup-python@v5
+        with:
+          python-version: ${{ env.PYTHON_VERSION }}
+      - run: pip install boto3
+      - uses: taiki-e/install-action@v2
+        with:
+          tool: nextest@0.9.82
+      - name: Install Package Dependencies
+        run: make install
+      - name: Start Integration Test Environment
+        run: make integration-env-up-ci
+      - name: Wait for Test Artifacts
+        uses: kachick/wait-other-jobs@v3.6.0
+        timeout-minutes: 5
```

Comment:
Why is necessary? Oh because of the `--detach` the above step finishes nearly immediately? Before we would get lucky, maybe?

Reply:
Well before we ran integration tests serially after unit tests.

You'll notice that "Run Unit Tests" has a direct dependency on "Build Test Aetifacts". That works well because the latter doesn't really take any startup time.

For integration tests, we need to start the integration env, i.e. pull docker images, start containers, and build `emily-server`, which takes ~1m.

So, instead of having that 1m come _after_ the tests are built, we immediately start setting up the environment in parallel to the test artifacts being built, which also takes ~1m.

So this action waits for the artifacts to be ready (the nextest archives) if the environment would become ready before the artifacts are available for download.

Hope that makes sense?

Reply:
Yup, that makes sense. Before this change, we would rely on the fact that it takes time to run unit tests.

Reply:
> Before this change, we would rely on the fact that it takes time to run unit tests before running our integration tests.

Well, before this change we would actually do the following, serially:

1. Run `clippy` (build 1, no `--tests` flag)
2. Run unit tests (build 2, `--tests` flag)
3. Setup test environment (build 3, `emily-server` with `--bin` flag + docker pulls)
4. Run integration tests (build 4, `--tests` flag with `integration-tests` feature)

So this is all about doing as much of the above in parallel as possible, and avoiding re-compiling things where possible.